mtosoftware/FinalEngine
# ✨ Implement Component Property Management UI in Entity Inspector

# Description

- Added missing `PropertyViewModels` and views such as:
  - `DoublePropertyView`
  - `FloatPropertyView`
  - `IntPropertyView`
  - `StringPropertyView`
  - `Vector2PropertyView`
  - etc
- Updated the UI to use the view models created in #260.

Fixes #261 

## Dependencies

There are no new dependencies introduced in this PR.

## Type of change


## How Has This Been Tested?

I have provided unit tests to accommodate my changes. I ensured the application works as expected and that I can create an entity, display multiple components and modify their values.

**Test Configuration**:
* Operating System: Windows 10 Home
* Hardware: Intel i7-6700HQ, 16GB, GTX 950M
* Toolchain: VS Community 2022 17.5.4

https://github.com/softwareantics/FinalEngine/assets/50978201/52c918e3-1328-4ba7-96e9-eb1904aec25c

## Steps to Test

_You can follow these steps as a guideline but should take precaution to ensure that the feature functions as expected_.

1. Create a new `TestComponent` class that inherits from `IEntityComponent`. An example is shown below. To make testing easier you can temporarily disable the `TreatWarningsAsErrors` configuration so that you don't have to provide documentation just for testing purposes.

```csharp
public sealed class TestComponent : IEntityComponent
{
    public TestComponent()
    {
        this.IsTested = true;
    }

    //// This should not be shown in the editor.
    [Browsable(false)]
    public bool IsNotBrowsable { get; set; }

    //// This should not be shown in the editor.
    public bool IsNotPublicSetter { get; }

    public bool IsTested { get; set; }

    public double TestDouble { get; set; }

    public float TestFloat { get; set; }

    public Vector4 TestVector4 { get; set; }

    public Vector2 Vector2Test { get; set; }

    public Vector3 Vector3Test { get; set; }
}
```

2. Go to the `Scene.AddEntity` method and when an entity is added ensure that the new `TestComponent` is added to the new `Entity`.

```csharp
entity.AddComponent<TestComponent>();
```

3. Start up _FinalEngine.Editor.Desktop_ on your machine.
4. Go to _Scene > Create Entity_
5. Provide a name for the _Entity_ and click _Create_.
6. Click on the new _Entity_ in the _Scene Hierarchy_, this should change the `Properties` tool window to the "Entity Inspector".
7. Ensure that all `public` properties that have both public `get` and `set` accessors appear in the "Entity Inspector.
8. Ensure that properties within the `TestComponent` that are marked as `[Browsable(false)]` are hidden from the inspector.
9. Ensure that properties within the `TestComponent` that have only a `get` accessor are hidden from the editor.

## Known Issues to be Fixed

The issues listed below have been added to the [backlog](https://github.com/users/softwareantics/projects/2/views/1).

- At the moment, the system doesn't have the capability to handle nested properties and components. It's currently designed to work with simpler data types and vectors.
- Any attributes you apply to components won't be cross-checked and enforced in the Entity Inspector, though this might change in the future.
- The textboxes designed for float and double properties unfortunately don't allow the use of decimal points (e.g., 3.14).
- There are some limitations with the tab indexing feature, which may not work as expected for swiftly navigating through component properties.
- Regrettably, when you pick an entity, any components that were previously hidden won't retain that state.
- Adding or removing components directly within the entity inspector isn't feasible for now, although this feature is in the pipeline.
- Lastly, documentation for this feature is not very nice, I am planning to go over the editor and update all documentation later this week.

## Checklist:
